tests/integration/integration-test.py

```diff
@@ -79,93 +85,166 @@ def test_get_notes_query_by_student_id(self, endpoint='/notes'):
             for resource in response.json()['data']:
                 actual_source = resource['attributes']['source']
                 self.assertEqual(actual_source, source)
-
-    def query_sort_keys(self, student_id, endpoint='/notes'):
-        sort_keys = ['lastModified', 'source', 'permissions'] # , 'contextType']
-        for sort_key in sort_keys:
-            params = {'studentId': student_id, 'sortKey': sort_key}
-            response = utils.make_request(self, endpoint, 200, params=params)
-            note_schema = utils.get_resource_schema(self, 'NoteResource')
-            utils.check_schema(self, response, note_schema)
+        # Check more than one source at a time
+        if len(sources) > 0:
+            array_len = random.randint(1, len(sources))
+            rand_sources = random.sample(sources, array_len)
+            sources = ",".join(rand_sources)
+            params = {'studentId': student_id, 'sources': sources}
+            response = self.get_response(params)
             for resource in response.json()['data']:
-                sorted_element = resource['attributes'][sort_key]
+                    actual_source = resource['attributes']['source']
+                    self.assertIn(actual_source, sources)
 
-    # def isSorted(x, key = lambda x: x): 
-    #     return all([key(x[i]) <= key(x[i + 1]) for i in xrange(len(x) - 1)])
+        # invalid sources tests returns 400
+        invalid_sources = [" ", "advisorportal", "random", "degreeworks"]
+        invalid_source = random.choice(invalid_sources)
+        params = {'studentId': student_id, 'sources': invalid_source}
+        response = utils.make_request(self, endpoint, 400, params=params)
+        schema = utils.get_resource_schema(self, 'Error')
+        utils.check_schema(self, response, schema)
+
+    def query_context_type(self, student_id, context_types, endpoint='/notes'):
+        for context_type in context_types:
+            params = {'studentId': student_id, 'contextTypes': context_type}
+            response = self.get_response(params)
+            # Validating the contextTypes requested
+            # is the same contextTypes received
+            for resource in response.json()['data']:
+                context = resource['attributes']['context']
+                actual_context_type = context['contextType']
+                self.assertEqual(actual_context_type, context_type)
+        # Check more than one contextType at a time
+        array_len = random.randint(0, len(context_types))
+        rand_context_type = random.sample(context_types, array_len)
+        context_types = ",".join(rand_context_type)
+        params = {'studentId': student_id, 'contextTypes': context_types}
+        response = self.get_response(params)
+        for resource in response.json()['data']:
+            context = resource['attributes']['context']
+            actual_context_type = context['contextType']
+            self.assertIn(actual_context_type, context_types)
```

Comment:
This logic is repeated again. You should avoid repeating your code too many time either by combining similar logic or similar test cases together.

Reply:
similar to which lines?
in the loop before its using `assertEqual` and this one is `assertIn`
and I can't use the same logic i use in other functions because "contextType" is within "context" within "attributes" so it would be complicated a bit if i combine it with something else

Reply:
For example: line 209-211, 219-221, 237-240, etc.

I knew some of them use `assertEqual` and some of them use `assertIn`, and they are looking for different field from data set, but besides those, the logic is pretty much the same. Pseudo code:
```
for resource in fetched data:
    1. get actual data
    2. compare actual data with expected schema
```

I think you should be able to write a function with this shared logic but also be able to handle the variety.

Reply:
For example: line 209-211, 219-221, 237-240, etc.

I knew some of them use `assertEqual` and some of them use `assertIn`, and they are looking for different field from data set, but besides those, the logic is pretty much the same. Pseudo code:
```
for resource in fetched data:
    1. get actual data field and its schema
    2. compare actual data with expected schema
```

I think you should be able to write a function with this shared logic but also be capable to handle the variety.